### PR TITLE
Lint usage of `Debug`-based formatting

### LINKED ISSUE #634: Warn when using debug formatting in user-facing output
`format!` is the syntax extension used in `println!`, `panic!` and many other string formatting macros. It supports, among a lot of other cool stuff, two formatting modes: `Display` and `Debug` (optionally pretty-printed). This is described in the [fmt](http://doc.rust-lang.org/std/fmt/index.html) docs.

I propose a new lint which tries to warn when the Debug-mode (`{:?}` or `{:#?`}) is used in user-facing output.

Defining what 'user-facing' means is probably as much a challenge as writing the actual lint. At first, we could define it quite simple as: Everything that is _not_ in a test or a bench. Exceptions for logging macros could also be added.

Since this is only relevant to CLI applications, probably quite annoying when debugging, and could (at first) have many false-positives, I suggest making it Allow by default.

### PR DESCRIPTION
**Pull request by mcarton**
Fix #634.
Ref #461.
Ref  #442.

### CODE REVIEW

**Comment by mcarton** [comment]
Any comment on that one?

**Comment by llogiq** [comment]
Too much to look at from mobile; it'll have to wait until the afternoon.

File: src/consts.rs [hunk 1]
```diff
@@ -173,90 +171,6 @@ impl PartialOrd for Constant {
...
-    } else if b == 0x0a {
-        write!(fmt, "\\n")
-    } else if b == 0x0d {
-        write!(fmt, "\\r")
-    } else {
-        write!(fmt, "\\x{:02x}", b)
-    }
-}
-
-impl fmt::Display for Constant {
```
Comment (llogiq) [inline src/consts.rs]: Is there a specific reason for removing this (apart from it using Debug formatting to write out `Str`s)?

**Comment by llogiq** [comment]
Apart from one question, looks good to me.

File: src/consts.rs (hunk 1 above)
Comment (mcarton) [inline src/consts.rs]: It was used only once, only to print integers. After I refactored the loop lint to use `const_eval` it wasn’t used anymore and seemed of little use, if the user wrote `6*7` in an expression, they must have a reason not to have written `42`, a snippet should always be used to print the original expression.

**Comment by llogiq** [comment]
Thanks!

**Pull request merged**